### PR TITLE
enable customize activation functions in clip vision encoder

### PR DESCRIPTION
**Pull request by Gasoonjia**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?

This PR enables users to customize activation functions in clip vision encoder

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](../tests/torchtune) for any new functionality
- [x] update [docstrings](../docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [x] I have added an example to docs or docstrings;

### CODE REVIEW

File: torchtune/models/clip/_component_builders.py [hunk 1]
```diff
@@ -19,6 +19,7 @@ def clip_vision_encoder(
     output_cls_projection: bool = False,
     max_num_tiles: int = 4,
     in_channels: int = 3,
+    hidden_act: torch.nn.Module = torch.nn.SiLU(),
```
Comment (RdoubleA) [inline torchtune/models/clip/_component_builders.py:22]: nit: maybe call this `intermediate_activation` to be more clear which activation in the transformer it is changing?

File: torchtune/models/clip/_component_builders.py [hunk 2]
```diff
@@ -49,6 +50,7 @@ def clip_vision_encoder(
         max_num_tiles (int): The maximum number of tiles that can be processed. This is used to
             determine the size of the positional embeddings.
         in_channels (int): The number of image input channels.
+        hidden_act (torch.nn.Module): The activation function used in the transformer layers.
```
Comment (RdoubleA) [inline torchtune/models/clip/_component_builders.py:53]:
```suggestion
        hidden_act (torch.nn.Module): The activation function used in the intermediate layers in the transformer encoder
```

**Review by RdoubleA — COMMENTED** [review]
Hey @Gasoonjia, thanks for opening this PR. This generally makes sense to me. Out of curiosity, what is your goal for changing the activation function? Do you intend to retrain the clip encoder from scratch?

**Comment by felipemello1** [comment]
hey @Gasoonjia, you may be interested in our upcoming flamingo PR: https://github.com/pytorch/torchtune/pull/1357/files

feel free to ping me in workplace https://fb.workplace.com/profile.php?id=61556984579937

**Comment by Gasoonjia** [comment]
Hey @RdoubleA thanks for comment! I've updated the PR and pls take a look!
I'm working on leveraging torchtune's (tt) module to reproduce llava1.5 based on huggingface (hf) impl, and I realize that hf is using quickgelu instead of tt's default act (SiLU), so I'd like to have a way to customize the act func I use.

**Comment by felipemello1** [comment]
> reproduce llava1.5 based on huggingface

FYI, we had some work on llava transforms: https://github.com/pytorch/torchtune/pull/1057

**Comment by Gasoonjia** [comment]
Hey @felipemello1 glad to see you in the github! Yes I'm keeping my eye on your wonderful PR, and definetly need your help when I try to work on Flamingo! 
I noticed you are updating CLIP module in that PR, will there be huge update on that? Especially on modelling side?

**Comment by Gasoonjia** [comment]
> > reproduce llava1.5 based on huggingface
> 
> FYI, we had some work on llava transforms: #1057

Thanks for sharing! Will try to leverage your work!

**Review by RdoubleA — APPROVED** [review]
